src/s3ql/block_cache.py

```diff
@@ -70,7 +70,9 @@ def __init__(self, inode, blockno, filename, mode='w+b'):
         self.blockno = blockno
         self.last_write = 0
         self.pos = self.fh.tell()
-        self.size = os.fstat(self.fh.fileno()).st_size
+        # use allocation size instead of st_size
+        # to properly account for small files
+        self.size = os.fstat(self.fh.fileno()).st_blocks * 512
```

Comment:
This amounts to simply rounding up to the multiple of 512 blocks, but 512 rarely the blocksize that is used by the filesystem. IIRC ext4 defaults to 4k. Other filesystems may "inline" small files into the inode, so that this actually overestimates consumption. So I am not sure if this truly improves the situation.

Reply:
`st_blocks` is not measured in file system block sizes, but always in multiples of 512. [Python docs say so](https://docs.python.org/3.7/library/os.html#os.stat_result.st_blocks), as well as [stat(2) on Linux](http://man7.org/linux/man-pages/man2/stat.2.html) and [on FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=stat&sektion=2).

Reply:
As for overestimation, I think it's still better than unexpectedly running out of space and recovering from it by killing s3ql, growing the partition, fsck-ing it… (I was lucky, as all changes to the filesystem were limited to atime updates, so I was able to just flush contents and fsck from that. Without that, fsck actually fails in a variety of ways.)

Reply:
> `st_blocks` is not measured in file system block sizes, but always in multiples of 512. [Python docs say so](https://docs.python.org/3.7/library/os.html#os.stat_result.st_blocks), as well as [stat(2) on Linux](http://man7.org/linux/man-pages/man2/stat.2.html) and [on FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=stat&sektion=2).

Yes, that is exactly what I'm saying. You are simply rounding up to the next multiple of 512, no matter if this happens to be the block size used by the filesystem or not. So the result is not necessarily any better than simply taking the apparent size.

Reply:
If we want to be conservative, then rounding up to the next multiple of e.g. 4k is probably a better choice. But as I said in the bug, I think it may be better to let the user handle such issues through `--max-cache-entries`.

Reply:
> Yes, that is exactly what I'm saying. You are simply rounding up to the next multiple of 512, no matter if this happens to be the block size used by the filesystem or not.

You're missing the point. `st_blocks` is not about file system blocks, it is completely irrelevant here. In fact, you can test that yourself:

```
$ echo > foo
$ python -c "import os; print(os.stat('foo').st_blocks)"
8
```

Notice that it's 8, not 1. Again, it's not measured in file system blocks, it is measured in 512-byte blocks. File system block size is already accounted for in this interface.

Reply:
> > Yes, that is exactly what I'm saying. You are simply rounding up to the next multiple of 512, no matter if this happens to be the block size used by the filesystem or not.
> 
> You're missing the point. `st_blocks` is not about file system blocks, it is completely irrelevant here.

I don't think I am, since I am in full agreement with you on this :-).

> Notice that it's 8, not 1. Again, it's not measured in file system blocks, it is measured in 512-byte blocks. 

Exactly. So the effect of using st_blocks * 512 instead of st_size is to round up the actual size  the next multiple of 512. What I am trying to say is that this will not give you a better estimate of actual usage, since for that you need information about the filesystem block size (which, as you correctly say, fstat() does not give you).

Reply:
> So the effect of using st_blocks * 512 instead of st_size is to round up the actual size the next multiple of 512.

`st_blocks` does not measure actual file size, but its *allocaion size*, i.e. how much space it uses with all the overhead and other filesystem tricks. (That's why this field usually have remarks about sparse files, for instance.)

Again, see example above: `st_blocks` there is 8, i.e. 4096 bytes.